api/v1beta1/imagerepository_types.go

```diff
@@ -0,0 +1,146 @@
+	// +optional
+	Credentials ImageCredentials `json:"credentials,omitempty"`
+}
+
+// ImageParameters describes requested image repository configuration.
+type ImageParameters struct {
+	// Name of the image within configured Quay organization.
+	// If ommited, then defaults to "cr-namespace/cr-name".
+	// This field cannot be changed after the resource creation.
+	// +optional
+	// +kubebuilder:validation:Pattern="[a-z0-9][.a-z0-9_-]*(/[a-z0-9][.a-z0-9_-]*)*"
+	Name string `json:"name,omitempty"`
+
+	// Visibility defines whether the image is publicly visible.
+	// Allowed values are public and private.
+	// "public" is the default.
+	// +optional
+	Visibility ImageVisibility `json:"visibility,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=public;private
+type ImageVisibility string
+
+const (
+	ImageVisibilityPublic  ImageVisibility = "public"
+	ImageVisibilityPrivate ImageVisibility = "private"
+)
+
+type ImageCredentials struct {
+	// RegenerateToken defines a request to refresh image accessing credentials.
+	// Refreshes both, push and pull tokens.
+	// The field gets cleared after the refresh.
+	RegenerateToken *bool `json:"regenerate-token,omitempty"`
+}
+
+// ImageRepositoryStatus defines the observed state of ImageRepository
+type ImageRepositoryStatus struct {
+	// State shows if image repository could be used.
+	// "ready" means repository was created and usable,
+	// "failed" means that the image repository creation request failed.
+	State ImageRepositoryState `json:"state,omitempty"`
+
+	// Message shows error information for the request.
+	// It could contain non critical error, like failed to change image visibility,
+	// while the state is ready and image resitory could be used.
+	Message string `json:"message,omitempty"`
+
+	// Image describes actual state of the image repository.
+	Image ImageStatus `json:"image,omitempty"`
+
+	// Credentials contain information related to image repository credentials.
+	Credentials CredentialsStatus `json:"credentials,omitempty"`
+}
+
+type ImageRepositoryState string
+
+const (
+	ImageRepositoryStateReady  ImageRepositoryState = "ready"
+	ImageRepositoryStateFailed ImageRepositoryState = "failed"
+)
+
+// ImageStatus shows actual generated image repository parameters.
+type ImageStatus struct {
+	// URL is the full image repository url to push into / pull from.
+	URL string `json:"url,omitempty"`
+
+	// Visibility shows actual generated image repository visibility.
+	Visibility ImageVisibility `json:"visibility,omitempty"`
+}
+
+// CredentialsStatus shows information about generated image repository credentials.
+type CredentialsStatus struct {
+	// GenerationTime shows timestamp when the current credentials were generated.
+	GenerationTimestamp *metav1.Time `json:"generationTimestamp,omitempty"`
+
+	// PushSecretName holds name of the dockerconfig secret with credentials to push (and pull) into the generated repository.
+	PushSecretName string `json:"push-secret,omitempty"`
```

Comment:
Not a big deal, but would it be possible to specify this in the spec? 

At the moment the JVM Build Service assumes the secret is named `jvm-build-image-secrets`, and even if using the SPI will use the `LinkableSecretSpec` to make the SPI create this secret. 

It's not a big deal, as I could update the JVM build service to allow for alternate names.

Reply:
That's possible, but I'd prefer auto-generated names to avoid collisions and attempts to rename the secret by the user.

Reply:
ok, that is fine, I can deal with this on the JVM Build Service side.